### ISSUE TITLE
The problem about running one model twice and using one checkpoint in *.py

### ISSUE THREAD

**Issue by yayeoCddy** [Oneflow-Inc/oneflow#3542]
Recently, I adopt oneflow to train GAN with the perceptual loss. Particularly, this loss takes the extrated featuer from the pre-trained vgg16bn as input to calculate MSE loss. As a result, I meet with following difficulties:

 - 1) When training Discriminator(D) in GAN, I need to run twice D, like this:

```
	@flow.global_function(func_config)
	def train_discriminator(
		noise=flow.FixedTensorDef((self.batch_size, 3, 256, 256)),
		real_images=flow.FixedTensorDef((self.batch_size, 3, 256, 256)),
	):
	g_out = self.generator(noise, trainable=False)
    g_logits = self.discriminator(g_out, trainable=True)
	d_logits = self.discriminator(g_out, trainable=True, reuse=True)
	
	return ...
```
But it seems oneflow can not support running the forward propagation twice in a graph, because `variable` in oneflow can not use the same name. So I have to rewrite some basic operations(ops), such as conv, decovn and dense layers. Specifically, I set the different name (original_name and original_name+'_reuse') for op(layer) and set the same name for weight(variable) in op using `get_variable`, like this:
```
def conv2d(
    input,
    filters,
    size,
    name,
    strides=2,
    padding="same",
    trainable=True,
    reuse=False
):
    name_ = name if reuse == False else name + "_reuse"

    # (output_dim, k_h, k_w, input.shape[3]) if NHWC
    weight_shape = (filters, input.shape[1], size, size)
    weight = flow.get_variable(
        name + "-weight",
        shape=weight_shape,
        dtype=input.dtype,	initializer=flow.glorot_uniform_initializer(data_format="NCHW"),
        trainable=trainable,
    )

    output = flow.nn.compat_conv2d(
        input,
        weight,
        strides=[strides, strides],
        padding=padding,
        data_format="NCHW",
        name=name_,
    )
	
	return output

```
However, it can not seem to modify Batch_norm layer in this way. Therefore, I modify Batch_norm like this:
```
def _batch_norm(inputs, name, trainable=True, training=True):
    params_shape = [inputs.shape[1]]
    # Float32 required to avoid precision-loss when using fp16 input/output
    params_dtype = flow.float32 if inputs.dtype == flow.float16 else inputs.dtype
    if not flow.current_global_function_desc().IsTrainable() or not trainable:
        training = False
    with flow.scope.namespace(name):
        beta = flow.get_variable(
            name="beta",
            shape=params_shape,
            dtype=params_dtype,
            initializer=flow.zeros_initializer(),
            trainable=trainable,
            distribute=distribute_util.broadcast(),
        )
        gamma = flow.get_variable(
            name="gamma",
            shape=params_shape,
            dtype=params_dtype,
            initializer=flow.ones_initializer(),
            trainable=trainable,
            distribute=distribute_util.broadcast(),
        )
        moving_mean = flow.get_variable(
            name="moving_mean",
            shape=params_shape,
            dtype=params_dtype,
            initializer=flow.zeros_initializer(),
            trainable=False,
            distribute=distribute_util.broadcast(),
        )
        moving_variance = flow.get_variable(
            name="moving_variance",
            shape=params_shape,
            dtype=params_dtype,
            initializer=flow.ones_initializer(),
            trainable=False,
            distribute=distribute_util.broadcast(),
        )
    builder = (
        flow.user_op_builder(name)
        .Op("normalization")
        .Input("x", [inputs])
        .Input("moving_mean", [moving_mean])
        .Input("moving_variance", [moving_variance])
        .Input("gamma", [gamma])
        .Input("beta", [beta])
        .Output("y")
        .Attr("axis", 1)
        .Attr("epsilon", 1.001e-5)
        .Attr("training", training)
        .Attr("momentum", 0.997)
    )
    if trainable and training:
        builder = builder.Output("mean").Output("inv_variance")
    return builder.Build().InferAndTryRun().RemoteBlobList()[0]

# def batchnorm(input, name, axis=1, reuse=False, trainable=True):
#     name_ = name if reuse == False else name + "_reuse"
#     print('name',name)
#     print('name_',name_)
#     return flow.layers.batch_normalization(input, axis=axis,name=name, name_=name_,reuse=reuse , trainable=trainable)

def batchnorm(input, name, axis=1, reuse=False, trainable=True):
    # print('use our own bN')
    name =  name+'_reuse' if reuse else name
    return _batch_norm(input,name,trainable=trainable)
```

  But these setups are too complicated. 

  Hence, I should how to adopt oneflow to achieve the goal running twice D in one training config ? 

 - 2) When using the perceptual loss, I need to run twice vgg, like this:
   
   	```
	def perceptual_loss(fake, real, weight=1.0):
		fake_feature = self.vgg16bn(fake, trainable=False)
		real_feature = self.vgg16bn(real, trainable=False, reuse=True)

		return self.mseloss(fake_feature, real_feature, weight=weight)
```
   This is similar to the case above. In practice, I have to copy and rename the finetuned variable (name, name_'reuse') to run twice frozen vgg16bn.
   
 - 3) In summary, I have Generator, Discriminator and vgg in training process. The first two need to be trained, and last one needs to be frozen. Now, oneflow seems to use one `check_point = flow.train.CheckPoint()`. Therefore, it has to init, load and save simultaneously 3 models using `check_point.init()`, `check_point.load(vgg_path)`, `check_point.save(GAN_path)`. I meet with much WARNNING when loading vgg. It is because vgg is originally trained using SGD and Adam is used in my training process. When saving GAN, it saves trained GAN and frozen vgg in one path. 
    Though I could peel different models by naming them differently, it is so tanglesome.

**Comment by JamiePlur** [comment]
Thank you for your attention. The followings may help:
1.  For now, it is required to rewrite the layer of a custom reusable operation. But, we will simplify the reuse mechanism in the near future
2. We are now refactoring the checkpoint module to support a broader range of tasks which includes the need for selectively saving models.
3. Oneflow does not support the reuse of moving_mean and moving_variance in BN temporarily. It will not affect accuracy according to our tests.
